src/azul/types.py

```diff
@@ -158,6 +164,60 @@ def f(t):
     return tuple(set(f(t)))
 
 
+def get_generic_type_params(cls: type[Generic],
+                            *required_types: type
+                            ) -> list[Union[type, TypeVar]]:
+    """
+    Inspect and validate the type parameters of a subclass of `typing.Generic`.
+
+    Parameters that are bound at the class level are returned as types. Unbound
+    parameters are returned as `TypeVar`s. `*required_types` can be used to
+    assert the superclasses of a contiguous prefix of bound parameters.
+
+    >>> from typing import Generic, TypeVar
+    >>> S, T = TypeVar(name='S'), TypeVar(name='T')
+    >>> class A(Generic[S, T]):
+    ...     pass
+    >>> class B(A[int, T]):
+    ...     pass
+
+    >>> get_generic_type_params(A)
+    [~S, ~T]
+
+    >>> get_generic_type_params(A, int)
+    Traceback (most recent call last):
+    ...
+    TypeError: issubclass() arg 1 must be a class
+
+    >>> get_generic_type_params(B)
+    [<class 'int'>, ~T]
+
+    >>> get_generic_type_params(B, int)
+    [<class 'int'>, ~T]
+
+    >>> get_generic_type_params(B, str)
+    Traceback (most recent call last):
+    ...
+    AssertionError: (<class 'int'>, <class 'str'>)
+
+    >>> get_generic_type_params(B, int, int)
+    Traceback (most recent call last):
+    ...
+    TypeError: issubclass() arg 1 must be a class
+
+    >>> get_generic_type_params(B, int, int, int)
+    Traceback (most recent call last):
+    ...
+    AssertionError: 2
+    """
+    base_cls = one(getattr(cls, '__orig_bases__'))
+    types = get_args(base_cls)
+    assert len(required_types) <= len(types), len(types)
```

Comment:
```suggestion
    assert len(required_types) == len(types), len(types)
```

This is just leaking an implementation detail i.e. the fact that zip works that way. Otherwise, it's arbitrary. There is no use case for just checking some of the parameters. And if it were to become a use case, I don't think there is anything special about the first N parameters.